pkg/operator/controller/crl/crl_configmap.go

```diff
@@ -0,0 +1,295 @@
+			} else {
+				subjectKeyIds = append(subjectKeyIds, subjectKeyId)
+				continue
+			}
+		}
+		log.Info("retrieving certificate revocation list", "subject key identifier", subjectKeyId)
+		if crl, err := getCRL(cert.CRLDistributionPoints); err != nil {
+			// Creating or updating the configmap with incomplete
+			// data would compromise security by potentially
+			// permitting revoked certificates.
+			return false, nil, fmt.Errorf("failed to get certificate revocation list for certificate key %s: %w", subjectKeyId, err)
+		} else {
+			crls[subjectKeyId] = crl
+			subjectKeyIds = append(subjectKeyIds, subjectKeyId)
+		}
+	}
+
+	if len(subjectKeyIds) == 0 {
+		return false, nil, nil
+	}
+
+	buf := &bytes.Buffer{}
+	for _, subjectKeyId := range subjectKeyIds {
+		asn1Data, err := asn1.Marshal(*crls[subjectKeyId])
+		if err != nil {
+			return false, nil, fmt.Errorf("failed to encode ASN.1 for CRL for certificate key %s: %w", subjectKeyId, err)
+		}
+		block := &pem.Block{
+			Type:  "X509 CRL",
+			Bytes: asn1Data,
+		}
+		if err := pem.Encode(buf, block); err != nil {
+			return false, nil, fmt.Errorf("failed to encode PEM for CRL for certificate key %s: %w", subjectKeyId, err)
+		}
+	}
+	crlData := buf.String()
+
+	crlConfigmapName := controller.CRLConfigMapName(ic)
+	crlConfigmap := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      crlConfigmapName.Name,
+			Namespace: crlConfigmapName.Namespace,
+		},
+		Data: map[string]string{
+			"crl.pem": crlData,
+		},
+	}
+	crlConfigmap.SetOwnerReferences([]metav1.OwnerReference{ownerRef})
+
+	return true, &crlConfigmap, nil
+}
+
+// getCRL gets a certificate revocation list using the provided distribution
+// points and returns the certificate list.
+func getCRL(distributionPoints []string) (*pkix.CertificateList, error) {
+	var errs []error
+	for _, distributionPoint := range distributionPoints {
+		// The distribution point is typically a URL with the "http"
+		// scheme.  "https" is generally not used because the
+		// certificate list is signed, and because using TLS to get the
+		// certificate list could introduce a circular dependency
+		// (cannot use TLS without the revocation list, and cannot get
+		// the revocation list without using TLS).
+		//
+		// TODO Support ldap.
+		if !strings.HasPrefix(distributionPoint, "http:") {
+			errs = append(errs, fmt.Errorf("unsupported distribution point type: %s", distributionPoint))
+			continue
+		}
+		log.Info("retrieving CRL distribution point", "distribution point", distributionPoint)
+		resp, err := http.Get(distributionPoint)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error getting %q: %w", distributionPoint, err))
+			continue
+		}
+		bytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error reading response from %q: %w", distributionPoint, err))
+			continue
+		}
+		crl, err := x509.ParseCRL(bytes)
```

Comment:
Close the response body; `defer resp.Body.Close()`.

Reply:
That is very important!  Fixed.  Thanks!

Because this is in a loop, I am avoiding use of `defer` and instead doing this:
```go
		bytes, err := ioutil.ReadAll(resp.Body)
		// Close the body immediately rather than using defer because
		// this is inside a loop.
		err2 := resp.Body.Close()
		if err2 != nil {
			errs = append(errs, fmt.Errorf("error closing response body from %q: %w", distributionPoint, err2))
		}
		// Now check the error value from reading the body.
		if err != nil {
```
If you prefer, I could wrap the whole thing in a `defer func() { ... }()`, or refactor the loop body into a separate function, so that the deferred close happens before the next URL is tried.  Or maybe I'm overthinking this.  What do you think?

Reply:
You can't call `Close()` without first checking for an error because `Body` will be `nil` if there was an error. If the loop terminates quickly I'd wrap it in a defer. If not, then I'd explicitly close it after a successful read. (Not sure I'd error on close failing.)

Reply:
> You can't call `Close()` without first checking for an error because `Body` will be `nil` if there was an error.

I thought I could.  Wouldn't using `defer resp.Body.Close()` cause the function to call `resp.Body.Close()` irrespective of whether there were an error on the read?

Reply:
FWIW, here's an example from https://pkg.go.dev/net/http:
```go
resp, err := http.Get("http://example.com/")
if err != nil {
	// handle error
}
defer resp.Body.Close()
body, err := io.ReadAll(resp.Body)
```
We have similar code in the canary controller.  The deferred function will be called even if `io.ReadAll` returns an error.

Reply:
Ah, sorry. I was incorrectly focusing on the read and not the get.

Reply:
I went ahead and refactored the HTTP code into a separate function.  The overall structure is nicer now too, with a `switch` statement for the different distribution point types (although currently only HTTP is supported).